src/Raven.Server/Rachis/Follower.cs

```diff
@@ -467,6 +467,33 @@ private void NegotiateWithLeader(TransactionOperationContext context, LogLengthN
 
                 context.Transaction.Commit();
             }
+
+            _engine.Timeout.Defer(_connection.Source);
+            _debugRecorder.Record("Invoking StateMachine.SnapshotInstalled");
+
+            // notify the state machine, we do this in an async manner, and start
+            // the operator in a separate thread to avoid timeouts while this is
+            // going on
+
+            var task = Task.Run(() => _engine.SnapshotInstalledAsync(snapshot.LastIncludedIndex));
+
+            var sp = Stopwatch.StartNew();
+
+            var timeToWait = (int)(_engine.ElectionTimeout.TotalMilliseconds / 4);
+
+            while (task.Wait(timeToWait) == false)
+            {
+                // this may take a while, so we let the other side know that
+                // we are still processing, and we reset our own timer while
+                // this is happening
+                MaybeNotifyLeaderThatWeAreStillAlive(context, sp);
+            }
+
+            _debugRecorder.Record("Done with StateMachine.SnapshotInstalled");
+
+            // we might have moved from passive node, so we need to start the timeout clock
```

Comment:
(writing from my phone)
Can you check what happens when an old v4.0 raft is getting this messages, I'm not sure if we should maybe bump the raft tcp protocol version for this or not.

Reply:
Is it related to my question in the above comment or separate one?

Anyway, `MaybeNotifyLeaderThatWeAreStillAlive` sends `InstallSnapshotResponse` message having `Done = false`. I've found the following piece of code which has some logic based on value of `Done` property:
https://github.com/ravendb/ravendb/blob/v4.0/src/Raven.Server/Rachis/FollowerAmbassador.cs#L518-L527
As it's not a new concept it doesn't seem to me to be a protocol change.

Reply:
@arekpalinski  When we get here, we aren't passive, because we are now part of a cluster.  This happened in the `InstallSnapshot` section. So if we were passive, now we aren't and we want to register the timeout handler.

@talweiss1982 This is the same behavior as we had previous. We essentially have the same logic, notifying the other side that we are pending. Not protocol change here.